sdk/java-sdk-spring/src/test/java/kalix/spring/testmodels/action/ActionsTestModels.java

```diff
@@ -218,4 +221,12 @@ public Flux<Effect<Message>> message(@RequestBody Flux<Message> messages) {
       return messages.map(msg -> effects().reply(msg));
     }
   }
+
+  public static class OnStartupHookAction extends Action {
+    @PostMapping("/message")
+    @Trigger(on = STARTUP, maxRetries = 2)
```

Comment:
I find it better to use `on` instead of `value` as it reads better. 

On the other hand, I feel that we are making a future-proof API without having a clear plan. At the moment, we don't have any reason to think that there will be another TriggerEvent type.

But the decision we are taking is forcing the user to choose it from a list of one item. 

From the point of view of the user's API, we force them to write `\@Trigger`, then type `on = `, then look up for the enum to pass. And that will be as such for the foreseeable future. While an annotation as `\@TriggerOnStartUp(maxReqrites = 2)` would be easier for the user to write (no enum to look after) and not give the impression there is an option for event trigger. 

Unless we have a very clear use case for other TriggerEvents.

Reply:
Although we don't have immediate plans for that, we already have some customer feedback wanting other events:
> This was previously reported and the discussion yielded the idea to support full lifecycle event triggering so not just service startup, shutdown, node failure, cluster member failure, shard rebalancing, or whatever makes sense from a Kalix point of view. In our view, modern cloud native applications need to be aware of the state of the resources they are using. It doesn’t need to be super-fine-grained however.

While I don't think all of those are reasonable, I do think there is a high probability of needing to expand this for more events.. some of the more obvious for me are "before readiness" and "on shutdown".

For that reason, I think it is better to have the user spending the extra few seconds but allow for future evolution rather than having them/us go through deprecation cycles. Or would you consider then having a different annotation for each event?

Reply:
Having specialized annotations can be an option as well, but I was not thinking about that at all, because I have the feel that we will only have `OnStartUp` anyway. 

But fine. I just wanted to raise the question. Actually, I should have started this debate when we were working on the protocol, but it only occurred to me now.

Reply:
What if it was `Trigger.OnStartup`? Would it make a difference in your view?

I think we are still on time to discuss it btw.. I mean, this is for code-first, we are not limited by what we did at the protocol-first one anyway.

Reply:
Hmm, I like this because it make it quite discoverable if we add other trigger types. It suffices to type `Trigger.` and get a list of all possible event triggers. And then we can add specific params on each of them. 

> I think we are still on time to discuss it btw.. I mean, this is for code-first, we are not limited by what we did at the protocol-first one anyway.

I see your point and I've done similar in the past. For code-first we should not try to reproduce the same as Protobuf, because Java annotations can be more expressive. So we can make it more convenient. 

But on the other hand, if we believe it's better to have specialized annotations, why shouldn't we have for the Protobuf? In that particular case, there is not restriction.

Reply:
Let me clarify the above with an example.

Protobuf has `eventing.in` and that receives the type name of the entity. And that's a String. 

In Java code-first, we could force some typing. That's why we have: `@Subscribe.EventSourcedEntity(MyEntity.class)`. So we could make something more much expressive and not stringly-typed. 

While here, we could have specialized annotations on both sides.

I'm fine if we don't do it in Protobuf and have `Trigger.OnStartup` for java.

Reply:
> But on the other hand, if we believe it's better to have specialized annotations, why shouldn't we have for the Protobuf? In that particular case, there is no restriction.

yeah, sure. I agree, if it makes sense for code-first it would probably make sense for protocol-first as well.
I just think it's one of those things that the user will do once every 6 months and never touch again...

I will update this to `Trigger.OnStartup` and then we can decide what to do with protocol first.

Reply:
Btw, the advantage of using specialized interfaces is that fact that we can define different properties depending on the event, the disadvantage being that we will probably repeat some of the fields (i.e. maxRetries). In any case, I think it's a good tradeoff.

Reply:
Refactored in https://github.com/lightbend/kalix-jvm-sdk/pull/1924/commits/2b666cb623d09efc3694e4ca977e0c1969d21988